Bring back naive implementation of FAR for target-typed new

jcouv opened the pull request:
This reverts commit 68086128b703585ee2f2aa06b932825839a532ef.

CyrusNajmabadi commented:
I thought we were going to go with an approach using a background incremental analyzer to build a special index for these.

jcouv commented:
@CyrusNajmabadi Yes, but we also discussed making progress by enabling the scenario and getting a sense of the perf problem. This PR would allow us to modify Roslyn with some target-typed new and get a feel. What do you think?

CyrusNajmabadi commented:
>  This PR would allow us to modify Roslyn with some target-typed new and get a feel. What do you think?

Got it.  For me, i thought we were saying that we woudl more do it as an experiment.  i.e. it's fine to have this is draft, and use it to get some perf numbers.  But i'm not sure i would want to preemptively merge things until we had that info.  Is that ok with you?

jcouv commented:
Re-opened PR following experiment we discussed.
In short:
I changed about half of object creations in CodeAnalysis.CSharp to use target-type new.
Then did a FindAllReferences on a few constructors. Also added some edits (adding a new type) in the mix.
The performance seems similar to invoking FindAllReferences on a popular method name (like `Equals`), which is what we’d expect.

jcouv commented:
@CyrusNajmabadi I rebased to resolve conflicts. Feel free to take another look when you have time.

CyrusNajmabadi commented on src/EditorFeatures/Test2/FindReferences/FindReferencesTests.ConstructorSymbols.vb:
```diff
@@ -818,5 +818,81 @@ public class A
+</Workspace>
+            Await TestAPIAndFeature(input, kind, host)
+        End Function
+
```
```suggestion
```

CyrusNajmabadi commented on src/EditorFeatures/Test2/FindReferences/FindReferencesTests.ConstructorSymbols.vb:
```diff
@@ -818,5 +818,81 @@ public class A
+        End Function
+
     End Class
 End Namespace
```
we should have some additional tests.  for example, we should verify that reorder parameters (and all our other sig-change features) doesn't crash.  if it does the right ting, great!  if it doesn't open bug.

@dpoeschl  can give you a full list.

jcouv replied:
ChangeSignature indeed has some problems. Opened https://github.com/dotnet/roslyn/issues/44126

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs:
```diff
@@ -564,6 +567,33 @@ static bool IsRelevantDocument(SyntaxTreeIndex syntaxTreeInfo)
+                    var symbolUsageInfo = GetSymbolUsageInfo(node, semanticModel, syntaxFacts, semanticFacts, cancellationToken);
+
+                    locations.Add(new FinderLocation(node, new ReferenceLocation(
+                        document, alias: null, location, isImplicit: true, symbolUsageInfo, GetAdditionalFindUsagesProperties(node, semanticModel, syntaxFacts), CandidateReason.None)));
```
does `GetAdditionalFindUsagesProperties(node, semanticModel, syntaxFacts)` work on these?

jcouv replied:
Strangely, I could not find a test that is affected by commenting out logic in `GetAdditionalFindUsagesProperties(node, semanticModel, syntaxFacts)`.
It works though (tested manually) and also reviewed what compiler APIs it calls.

![image](https://user-images.githubusercontent.com/12466233/81513341-bb3cb600-92dc-11ea-80a1-0ca16e3df97a.png)

CyrusNajmabadi replied:
👍

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:
```diff
@@ -40,12 +40,20 @@ protected override bool CanFind(IMethodSymbol symbol)
         }
 
+        private bool IsConstructor(IMethodSymbol methodSymbol)
+            => methodSymbol.MethodKind == MethodKind.Constructor;
```
1. don't know if this method feels warranted :)
2. static

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/SyntaxKindEx.cs:
```diff
@@ -15,6 +15,8 @@ namespace Microsoft.CodeAnalysis.CSharp.Formatting
     [SuppressMessage("CodeQuality", "IDE0051:Remove unused private members", Justification = "Private constants are used for static assertions.")]
     internal static class SyntaxKindEx
     {
+        public const SyntaxKind ImplicitObjectCreationExpression = (SyntaxKind)8659;
```
can you add teh same overflow check below in `#!CODE_STYLE`?

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpSemanticFactsService.cs:
```diff
@@ -401,5 +401,16 @@ private SymbolInfo GetSymbolInfo(SemanticModel semanticModel, SyntaxNode node, S
+            if (node.Kind() == Formatting.SyntaxKindEx.ImplicitObjectCreationExpression)
+            {
+                var info = semanticModel.GetSymbolInfo(node);
+                return (IMethodSymbol)info.Symbol;
```
is this what we do for normal constructors?  i.e. this is ignoring candidates in the case of errors?  if that is what we do normally, grat.  but if we normally use any/all of the symbols we get back for a normal constructor, then we should follow that here.

CyrusNajmabadi replied:
also, i don't think this method is needed.  it is necessary when we'll need some sort of funky semantic model call to get specific information (like foreach-info), but for FAR, it could just call .GetSymbolInfo(node) since it already knows it's an implici-obje-expr.

jcouv replied:
I had not thought of that. As far as I understand, the logic for FAR on constructors will only consider non-error symbols (uses `GetSymbolInfo(...).Symbol` only, ignoring candidates):

```
        protected static Func<SyntaxNode, SemanticModel, (bool matched, CandidateReason reason)> GetStandardSymbolsNodeMatchFunction(
            ISymbol searchSymbol, Solution solution, CancellationToken cancellationToken)
        {
            return (node, model) =>
            {
                var symbolInfoToMatch = FindReferenceCache.GetSymbolInfo(model, node, cancellationToken);

                var symbolToMatch = symbolInfoToMatch.Symbol;
                var symbolToMatchCompilation = model.Compilation;

                if (SymbolFinder.OriginalSymbolsMatch(searchSymbol, symbolInfoToMatch.Symbol, solution, null, symbolToMatchCompilation, cancellationToken))
                {
                    return (matched: true, CandidateReason.None);
                }
                else if (symbolInfoToMatch.CandidateSymbols.Any(s => SymbolFinder.OriginalSymbolsMatch(searchSymbol, s, solution, null, symbolToMatchCompilation, cancellationToken)))
                {
                    return (matched: true, symbolInfoToMatch.CandidateReason);
                }
                else
                {
                    return (matched: false, CandidateReason.None);
                }
            };
        }

```

---
In reply to: [422305225](https://github.com/dotnet/roslyn/pull/43645#discussion_r422305225) [](ancestors = 422305225)

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:
```diff
@@ -96,8 +101,9 @@ protected override bool CanFind(IMethodSymbol symbol)
             var ordinaryRefs = await FindOrdinaryReferencesAsync(symbol, document, semanticModel, findParentNode, cancellationToken).ConfigureAwait(false);
             var attributeRefs = await FindAttributeReferencesAsync(symbol, document, semanticModel, cancellationToken).ConfigureAwait(false);
             var predefinedTypeRefs = await FindPredefinedTypeReferencesAsync(symbol, document, semanticModel, cancellationToken).ConfigureAwait(false);
+            var implicitOjbectCreationMatches = await FindReferencesInImplicitObjectCreationExpressionAsync(symbol, document, semanticModel, cancellationToken).ConfigureAwait(false);
```
Ojbect

CyrusNajmabadi commented on src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/ConstructorSymbolReferenceFinder.cs:
```diff
@@ -96,8 +101,9 @@ protected override bool CanFind(IMethodSymbol symbol)
+            var implicitOjbectCreationMatches = await FindReferencesInImplicitObjectCreationExpressionAsync(symbol, document, semanticModel, cancellationToken).ConfigureAwait(false);
 
-            return ordinaryRefs.Concat(attributeRefs).Concat(predefinedTypeRefs);
+            return ordinaryRefs.Concat(attributeRefs).Concat(predefinedTypeRefs).Concat(implicitOjbectCreationMatches);
```
aside, do we have a Concat that takes in multiple Immutable arrays.  it's a pity about all these copies.

jcouv replied:
We don't have such a Concat overload. Added
